File: hdf-converters/shared/go/vex/mapping_test.go
package vex

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestNormalizeStatus(t *testing.T) {
	t.Parallel()
	cases := map[string]Status{
		"not_affected":           StatusNotAffected,
		"known_not_affected":     StatusNotAffected,
		"false_positive":         StatusNotAffected,
		"affected":               StatusAffected,
		"known_affected":         StatusAffected,
		"exploitable":            StatusAffected,
		"fixed":                  StatusFixed,
		"first_fixed":            StatusFixed,
		"resolved":               StatusFixed,
		"resolved_with_pedigree": StatusFixed,
		"under_investigation":    StatusUnderInvestigation,
		"in_triage":              StatusUnderInvestigation,
		"  NOT_AFFECTED  ":       StatusNotAffected,
	}
	for raw, want := range cases {
		got, ok := NormalizeStatus(raw)
		assert.Truef(t, ok, "expected %q to normalize", raw)
		assert.Equalf(t, want, got, "for %q", raw)
	}

	for _, unknown := range []string{"", "garbage", "recommended", "last_affected"} {
		_, ok := NormalizeStatus(unknown)
		assert.Falsef(t, ok, "expected %q to be unknown", unknown)
	}
}

func TestNormalizeJustification(t *testing.T) {
	t.Parallel()
	cases := map[string]hdf.Justification{
		"component_not_present":                             hdf.ComponentNotPresent,
		"code_not_present":                                  hdf.ComponentNotPresent,
		"vulnerable_code_not_present":                       hdf.VulnerableCodeNotPresent,
		"vulnerable_code_not_in_execute_path":               hdf.VulnerableCodeNotInExecutePath,
		"code_not_reachable":                                hdf.VulnerableCodeNotInExecutePath,
		"vulnerable_code_cannot_be_controlled_by_adversary": hdf.VulnerableCodeCannotBeControlledByAdversary,
		"inline_mitigations_already_exist":                  hdf.InlineMitigationsAlreadyExist,
		"protected_by_mitigating_control":                   hdf.InlineMitigationsAlreadyExist,
		// CycloneDX-specific reachability values now in the HDF enum.
		"requires_configuration": hdf.RequiresConfiguration,
		"requires_dependency":    hdf.RequiresDependency,
		"requires_environment":   hdf.RequiresEnvironment,
		"protected_by_compiler":  hdf.ProtectedByCompiler,
		"protected_at_runtime":   hdf.ProtectedAtRuntime,
		"protected_at_perimeter": hdf.ProtectedAtPerimeter,
	}
	for raw, want := range cases {
		got, ok := NormalizeJustification(raw)
		assert.Truef(t, ok, "expected %q to normalize", raw)
		assert.Equalf(t, want, got, "for %q", raw)
	}

	for _, unknown := range []string{"", "garbage", "some_future_ecosystem_label"} {
		_, ok := NormalizeJustification(unknown)
		assert.Falsef(t, ok, "expected %q to be unknown — future ecosystems should extend the enum, not rely on reason-field passthrough", unknown)
	}
}

func TestImportTargetFor(t *testing.T) {
	t.Parallel()

	target, ok := ImportTargetFor(StatusNotAffected)
	assert.True(t, ok)
	assert.Equal(t, hdf.FalsePositive, target.OverrideType)
	assert.NotNil(t, target.Status)
	assert.Equal(t, hdf.Passed, *target.Status)
	assert.True(t, target.SetJustification)

	target, ok = ImportTargetFor(StatusFixed)
	assert.True(t, ok)
	assert.Equal(t, hdf.Poam, target.OverrideType)
	require.NotNil(t, target.Status, "POA&M schema branch requires status; fixed pins it to failed pending re-scan")
	assert.Equal(t, hdf.Failed, *target.Status)
	assert.NotEmpty(t, target.POAMActionTemplate)

	_, ok = ImportTargetFor(StatusAffected)
	assert.False(t, ok, "affected is informational; consumer creates amendment later if they act")

	_, ok = ImportTargetFor(StatusUnderInvestigation)
	assert.False(t, ok, "under_investigation is informational")
}

func TestExportStatusFor(t *testing.T) {
	t.Parallel()

	_, ok := ExportStatusFor(nil, false, false)
	assert.False(t, ok, "no consumer action = no VEX statement")

	just := hdf.ComponentNotPresent
	got, ok := ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.FalsePositive, Justification: &just}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusNotAffected, got, "justification set always emits not_affected")

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.OverrideTypeWaiver}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusAffected, got, "waiver = consumer accepts risk on a real finding")

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.RiskAdjustment}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusAffected, got)

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.OperationalRequirement}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusAffected, got)

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.Poam}, true, true)
	assert.True(t, ok)
	assert.Equal(t, StatusFixed, got, "POAM closure requires BOTH all-milestones-complete AND chained-as-latest")

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.Poam}, true, false)
	assert.True(t, ok)
	assert.Equal(t, StatusAffected, got, "milestones done but no closure amendment chained = still in flight")

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.Poam}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusAffected, got)

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.FalsePositive}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusNotAffected, got)

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.Attestation}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusNotAffected, got)

	got, ok = ExportStatusFor(&hdf.StandaloneOverride{Type: hdf.Inherited}, false, false)
	assert.True(t, ok)
	assert.Equal(t, StatusNotAffected, got)
}

func TestSupplierEvidence(t *testing.T) {
	t.Parallel()

	ev := SupplierEvidence("https://example.com/openvex.json", "OpenVEX from Vendor X")
	assert.NotNil(t, ev)
	assert.Equal(t, hdf.URL, ev.Type)
	assert.Equal(t, "https://example.com/openvex.json", ev.Data)
	assert.NotNil(t, ev.Description)
	assert.Equal(t, "OpenVEX from Vendor X", *ev.Description)

	ev = SupplierEvidence("https://example.com/openvex.json", "")
	assert.NotNil(t, ev)
	assert.Equal(t, "Upstream VEX statement", *ev.Description, "fallback description when caller has none")

	assert.Nil(t, SupplierEvidence("", "anything"), "no URI = no evidence; don't fabricate")
	assert.Nil(t, SupplierEvidence("   ", "anything"))
}

func TestJustificationForCycloneDX(t *testing.T) {
	t.Parallel()

	// Long-form HDF -> short-form CycloneDX
	v, ok := JustificationForCycloneDX(hdf.ComponentNotPresent)
	require.True(t, ok)
	assert.Equal(t, "code_not_present", v)
	v, ok = JustificationForCycloneDX(hdf.VulnerableCodeNotInExecutePath)
	require.True(t, ok)
	assert.Equal(t, "code_not_reachable", v)
	v, ok = JustificationForCycloneDX(hdf.InlineMitigationsAlreadyExist)
	require.True(t, ok)
	assert.Equal(t, "protected_by_mitigating_control", v)

	// CycloneDX-specific values pass through unchanged
	for _, val := range []hdf.Justification{
		hdf.RequiresConfiguration, hdf.RequiresDependency, hdf.RequiresEnvironment,
		hdf.ProtectedByCompiler, hdf.ProtectedAtRuntime, hdf.ProtectedAtPerimeter,
	} {
		got, gotOK := JustificationForCycloneDX(val)
		require.Truef(t, gotOK, "%s should pass through", val)
		assert.Equal(t, string(val), got)
	}

	// HDF-only values have no CycloneDX equivalent
	_, ok = JustificationForCycloneDX(hdf.VulnerableCodeNotPresent)
	assert.False(t, ok, "vulnerable_code_not_present has no CycloneDX equivalent")
	_, ok = JustificationForCycloneDX(hdf.VulnerableCodeCannotBeControlledByAdversary)
	assert.False(t, ok, "vulnerable_code_cannot_be_controlled_by_adversary has no CycloneDX equivalent")
}

func strptr(s string) *string { return &s }

func TestSwapPurlVersion(t *testing.T) {
	t.Parallel()
	assert.Equal(t, "pkg:npm/abc@4.5?arch=x64", SwapPurlVersion("pkg:npm/abc@4.2?arch=x64", "4.5"))
	assert.Equal(t, "pkg:npm/abc@4.5", SwapPurlVersion("pkg:npm/abc", "4.5"))
	assert.Equal(t, "pkg:npm/abc@4.5?arch=x64", SwapPurlVersion("pkg:npm/abc?arch=x64", "4.5"))
}

func TestFixedPackageIdentifier(t *testing.T) {
	t.Parallel()
	id, ok := FixedPackageIdentifier(hdf.AffectedPackage{Purl: strptr("pkg:npm/abc@4.2"), FixedInVersion: strptr("4.5")})
	require.True(t, ok)
	assert.Equal(t, "pkg:npm/abc@4.5", id)

	id, ok = FixedPackageIdentifier(hdf.AffectedPackage{Name: strptr("abc"), FixedInVersion: strptr("4.5")})
	require.True(t, ok)
	assert.Equal(t, "abc@4.5", id)

	_, ok = FixedPackageIdentifier(hdf.AffectedPackage{Purl: strptr("pkg:npm/abc@4.2")})
	assert.False(t, ok)
	_, ok = FixedPackageIdentifier(hdf.AffectedPackage{Cpe: strptr("cpe:2.3:a:x:y:1:*:*:*:*:*:*:*"), FixedInVersion: strptr("4.5")})
	assert.False(t, ok)
}

func TestVersTypeFor(t *testing.T) {
	t.Parallel()
	eco := hdf.Ecosystem("Npm")
	typ, ok := VersTypeFor(hdf.AffectedPackage{Ecosystem: &eco})
	require.True(t, ok)
	assert.Equal(t, "npm", typ)

	typ, ok = VersTypeFor(hdf.AffectedPackage{Purl: strptr("pkg:RPM/openssl@1.1")})
	require.True(t, ok)
	assert.Equal(t, "rpm", typ)

	_, ok = VersTypeFor(hdf.AffectedPackage{Cpe: strptr("cpe:2.3:a:x:y:1:*:*:*:*:*:*:*")})
	assert.False(t, ok)
}

File: hdf-converters/shared/typescript/vex/mapping.test.ts
import { describe, expect, it } from 'vitest';
import {
  Justification,
  OverrideType,
  ResultStatus,
} from '@mitre/hdf-schema';
import {
  affectedPackageFromIdentifier,
  affectedPackageToIdentifier,
  affectedPackagesFromIdentifiers,
  exportStatusFor,
  fixedPackageIdentifier,
  importTargetFor,
  justificationForCycloneDX,
  normalizeJustification,
  normalizeStatus,
  supplierEvidence,
  swapPurlVersion,
  versTypeFor,
  VexStatus,
} from './mapping.js';

describe('normalizeStatus', () => {
  it.each([
    ['not_affected', VexStatus.NotAffected],
    ['known_not_affected', VexStatus.NotAffected],
    ['false_positive', VexStatus.NotAffected],
    ['affected', VexStatus.Affected],
    ['known_affected', VexStatus.Affected],
    ['exploitable', VexStatus.Affected],
    ['fixed', VexStatus.Fixed],
    ['first_fixed', VexStatus.Fixed],
    ['resolved', VexStatus.Fixed],
    ['resolved_with_pedigree', VexStatus.Fixed],
    ['under_investigation', VexStatus.UnderInvestigation],
    ['in_triage', VexStatus.UnderInvestigation],
    ['  NOT_AFFECTED  ', VexStatus.NotAffected],
  ])('maps %s', (raw, expected) => {
    expect(normalizeStatus(raw)).toBe(expected);
  });

  it.each(['', 'garbage', 'recommended', 'last_affected'])(
    'returns undefined for %s',
    (raw) => {
      expect(normalizeStatus(raw)).toBeUndefined();
    },
  );
});

describe('normalizeJustification', () => {
  it.each([
    ['component_not_present', Justification.ComponentNotPresent],
    ['code_not_present', Justification.ComponentNotPresent],
    ['vulnerable_code_not_present', Justification.VulnerableCodeNotPresent],
    [
      'vulnerable_code_not_in_execute_path',
      Justification.VulnerableCodeNotInExecutePath,
    ],
    ['code_not_reachable', Justification.VulnerableCodeNotInExecutePath],
    [
      'vulnerable_code_cannot_be_controlled_by_adversary',
      Justification.VulnerableCodeCannotBeControlledByAdversary,
    ],
    [
      'inline_mitigations_already_exist',
      Justification.InlineMitigationsAlreadyExist,
    ],
    [
      'protected_by_mitigating_control',
      Justification.InlineMitigationsAlreadyExist,
    ],
    // CycloneDX-specific reachability values now in the HDF enum.
    ['requires_configuration', Justification.RequiresConfiguration],
    ['requires_dependency', Justification.RequiresDependency],
    ['requires_environment', Justification.RequiresEnvironment],
    ['protected_by_compiler', Justification.ProtectedByCompiler],
    ['protected_at_runtime', Justification.ProtectedAtRuntime],
    ['protected_at_perimeter', Justification.ProtectedAtPerimeter],
  ])('maps %s', (raw, expected) => {
    expect(normalizeJustification(raw)).toBe(expected);
  });

  it.each(['', 'garbage', 'some_future_ecosystem_label'])(
    'returns undefined for %s (future ecosystems should extend the enum)',
    (raw) => {
      expect(normalizeJustification(raw)).toBeUndefined();
    },
  );
});

describe('importTargetFor', () => {
  it('not_affected → falsePositive + passed + justification', () => {
    const target = importTargetFor(VexStatus.NotAffected);
    expect(target).toBeDefined();
    expect(target!.overrideType).toBe(OverrideType.FalsePositive);
    expect(target!.status).toBe(ResultStatus.Passed);
    expect(target!.setJustification).toBe(true);
  });

  it('fixed → POAM pinned to failed pending re-scan (real system has not been verified)', () => {
    const target = importTargetFor(VexStatus.Fixed);
    expect(target).toBeDefined();
    expect(target!.overrideType).toBe(OverrideType.Poam);
    expect(target!.status).toBe(ResultStatus.Failed);
    expect(target!.poamActionTemplate).not.toBe('');
  });

  it('affected → no amendment (informational; consumer acts later)', () => {
    expect(importTargetFor(VexStatus.Affected)).toBeUndefined();
  });

  it('under_investigation → no amendment', () => {
    expect(importTargetFor(VexStatus.UnderInvestigation)).toBeUndefined();
  });
});

const baseOverride = (
  type: OverrideType,
  justification?: Justification,
) => ({
  appliedAt: new Date('2026-06-06'),
  appliedBy: { type: 'simple', identifier: 'test' },
  expiresAt: new Date('2027-06-06'),
  reason: 'test',
  requirementId: 'AC-1',
  type,
  justification,
});

describe('exportStatusFor', () => {
  it('returns undefined for no override (consumer has not acted)', () => {
    expect(exportStatusFor(undefined, false, false)).toBeUndefined();
  });

  it('justification set → not_affected (regardless of override type)', () => {
    const o = baseOverride(OverrideType.FalsePositive, Justification.ComponentNotPresent);
    expect(exportStatusFor(o as never, false, false)).toBe(VexStatus.NotAffected);
  });

  it('waiver → affected', () => {
    expect(
      exportStatusFor(baseOverride(OverrideType.Waiver) as never, false, false),
    ).toBe(VexStatus.Affected);
  });

  it('riskAdjustment → affected', () => {
    expect(
      exportStatusFor(
        baseOverride(OverrideType.RiskAdjustment) as never,
        false,
        false,
      ),
    ).toBe(VexStatus.Affected);
  });

  it('operationalRequirement → affected', () => {
    expect(
      exportStatusFor(
        baseOverride(OverrideType.OperationalRequirement) as never,
        false,
        false,
      ),
    ).toBe(VexStatus.Affected);
  });

  it('poam, milestones complete AND closure chained → fixed', () => {
    expect(
      exportStatusFor(baseOverride(OverrideType.Poam) as never, true, true),
    ).toBe(VexStatus.Fixed);
  });

  it('poam, milestones complete but NOT chained → affected', () => {
    expect(
      exportStatusFor(baseOverride(OverrideType.Poam) as never, true, false),
    ).toBe(VexStatus.Affected);
  });

  it('poam, milestones not complete → affected', () => {
    expect(
      exportStatusFor(baseOverride(OverrideType.Poam) as never, false, false),
    ).toBe(VexStatus.Affected);
  });

  it.each([OverrideType.FalsePositive, OverrideType.Attestation, OverrideType.Inherited])(
    '%s with no justification → not_affected',
    (type) => {
      expect(exportStatusFor(baseOverride(type) as never, false, false)).toBe(
        VexStatus.NotAffected,
      );
    },
  );
});

describe('supplierEvidence', () => {
  it('builds a URL evidence entry from a source URI', () => {
    const ev = supplierEvidence(
      'https://example.com/openvex.json',
      'OpenVEX from Vendor X',
    );
    expect(ev).toBeDefined();
    expect(ev!.type).toBe('url');
    expect(ev!.data).toBe('https://example.com/openvex.json');
    expect(ev!.description).toBe('OpenVEX from Vendor X');
  });

  it('falls back to default description when caller has none', () => {
    const ev = supplierEvidence('https://example.com/openvex.json');
    expect(ev!.description).toBe('Upstream VEX statement');
  });

  it('returns undefined for empty URI (no fabrication)', () => {
    expect(supplierEvidence('')).toBeUndefined();
    expect(supplierEvidence('   ')).toBeUndefined();
  });
});

describe('justificationForCycloneDX', () => {
  it.each([
    [Justification.ComponentNotPresent, 'code_not_present'],
    [Justification.VulnerableCodeNotInExecutePath, 'code_not_reachable'],
    [Justification.InlineMitigationsAlreadyExist, 'protected_by_mitigating_control'],
  ])('translates long-form %s to CycloneDX short-form', (hdfValue, cdxValue) => {
    expect(justificationForCycloneDX(hdfValue)).toBe(cdxValue);
  });

  it.each([
    Justification.RequiresConfiguration,
    Justification.RequiresDependency,
    Justification.RequiresEnvironment,
    Justification.ProtectedByCompiler,
    Justification.ProtectedAtRuntime,
    Justification.ProtectedAtPerimeter,
  ])('passes CycloneDX-specific value %s through unchanged', (v) => {
    expect(justificationForCycloneDX(v)).toBe(String(v));
  });

  it.each([
    Justification.VulnerableCodeNotPresent,
    Justification.VulnerableCodeCannotBeControlledByAdversary,
  ])('returns undefined for %s (no CycloneDX equivalent)', (v) => {
    expect(justificationForCycloneDX(v)).toBeUndefined();
  });
});

describe('affectedPackageFromIdentifier', () => {
  it('parses a purl into name+version+ecosystem', () => {
    const pkg = affectedPackageFromIdentifier('pkg:npm/lodash@4.17.20');
    expect(pkg).toEqual({
      purl: 'pkg:npm/lodash@4.17.20',
      name: 'lodash',
      version: '4.17.20',
      ecosystem: 'npm',
    });
  });

  it('falls back to ecosystem=generic for unknown purl type segments', () => {
    const pkg = affectedPackageFromIdentifier('pkg:apk/wolfi/git@2.39.0-r1');
    expect(pkg?.purl).toBe('pkg:apk/wolfi/git@2.39.0-r1');
    expect(pkg?.ecosystem).toBe('generic');
  });

  it('preserves a malformed pkg: identifier as purl-only', () => {
    // parsePurl returns null when the type segment is missing — preserve
    // the raw string verbatim rather than dropping the entry.
    const pkg = affectedPackageFromIdentifier('pkg:');
    expect(pkg).toEqual({ purl: 'pkg:' });
  });

  it('recognizes a CPE 2.3 identifier and emits cpe-only', () => {
    const pkg = affectedPackageFromIdentifier(
      'cpe:2.3:a:openssl:openssl:1.1.1k:*:*:*:*:*:*:*',
    );
    expect(pkg).toEqual({
      cpe: 'cpe:2.3:a:openssl:openssl:1.1.1k:*:*:*:*:*:*:*',
    });
  });

  it('returns undefined for opaque identifiers (schema forbids fabricating identity)', () => {
    expect(affectedPackageFromIdentifier('acme-internal-lib')).toBeUndefined();
    expect(affectedPackageFromIdentifier('')).toBeUndefined();
  });
});

describe('affectedPackagesFromIdentifiers', () => {
  it('dedupes by purl/cpe key and drops unresolvable entries', () => {
    const out = affectedPackagesFromIdentifiers([
      'pkg:npm/lodash@4.17.20',
      'pkg:npm/lodash@4.17.20', // dedup
      'opaque-string', // dropped
      '', // dropped
      'cpe:2.3:a:openssl:openssl:1.1.1k:*:*:*:*:*:*:*',
    ]);
    expect(out).toHaveLength(2);
    expect(out[0].purl).toBe('pkg:npm/lodash@4.17.20');
    expect(out[1].cpe).toBe('cpe:2.3:a:openssl:openssl:1.1.1k:*:*:*:*:*:*:*');
  });
});

describe('affectedPackageToIdentifier', () => {
  it('prefers purl over cpe and name+version', () => {
    expect(
      affectedPackageToIdentifier({
        purl: 'pkg:npm/x@1.0',
        cpe: 'cpe:2.3:a:vendor:x:1.0:*:*:*:*:*:*:*',
        name: 'x',
        version: '1.0',
      }),
    ).toBe('pkg:npm/x@1.0');
  });

  it('uses cpe when purl is absent', () => {
    expect(
      affectedPackageToIdentifier({
        cpe: 'cpe:2.3:a:vendor:x:1.0:*:*:*:*:*:*:*',
      }),
    ).toBe('cpe:2.3:a:vendor:x:1.0:*:*:*:*:*:*:*');
  });

  it('falls back to name@version when neither purl nor cpe is set', () => {
    expect(affectedPackageToIdentifier({ name: 'x', version: '1.0' })).toBe('x@1.0');
  });

  it('returns name alone when version is missing', () => {
    expect(affectedPackageToIdentifier({ name: 'x' })).toBe('x');
  });

  it('returns undefined for an empty AffectedPackage', () => {
    expect(affectedPackageToIdentifier({})).toBeUndefined();
  });
});

describe('swapPurlVersion', () => {
  it('swaps an existing @version, preserving the ?/# tail', () => {
    expect(swapPurlVersion('pkg:npm/abc@4.2?arch=x64', '4.5')).toBe('pkg:npm/abc@4.5?arch=x64');
  });
  it('inserts a version when the purl has none (no @)', () => {
    expect(swapPurlVersion('pkg:npm/abc', '4.5')).toBe('pkg:npm/abc@4.5');
    expect(swapPurlVersion('pkg:npm/abc?arch=x64', '4.5')).toBe('pkg:npm/abc@4.5?arch=x64');
  });
});

describe('fixedPackageIdentifier', () => {
  it('swaps the purl version when a purl is present', () => {
    expect(fixedPackageIdentifier({ purl: 'pkg:npm/abc@4.2', fixedInVersion: '4.5' })).toBe('pkg:npm/abc@4.5');
  });
  it('falls back to name@fixedInVersion without a purl', () => {
    expect(fixedPackageIdentifier({ name: 'abc', fixedInVersion: '4.5' })).toBe('abc@4.5');
  });
  it('returns undefined without a fixedInVersion or an anchor', () => {
    expect(fixedPackageIdentifier({ purl: 'pkg:npm/abc@4.2' })).toBeUndefined();
    expect(fixedPackageIdentifier({ cpe: 'cpe:2.3:a:x:y:1:*:*:*:*:*:*:*', fixedInVersion: '4.5' })).toBeUndefined();
  });
});

describe('versTypeFor', () => {
  it('prefers the ecosystem, lowercased', () => {
    expect(versTypeFor({ ecosystem: 'Npm' } as never)).toBe('npm');
  });
  it('derives the type from the purl when no ecosystem', () => {
    expect(versTypeFor({ purl: 'pkg:RPM/openssl@1.1' })).toBe('rpm');
  });
  it('returns undefined when neither ecosystem nor purl is set', () => {
    expect(versTypeFor({ cpe: 'cpe:2.3:a:x:y:1:*:*:*:*:*:*:*' })).toBeUndefined();
  });
});
